### DISCUSSION TITLE
Resize based on dpi and not pixel size

### DISCUSSION THREAD

**Discussion by hanssonrickard** [libvips/libvips#2188] [Q&A]
Hi.
What would be the smartest way of resizing an image based on a dpi value and not pixel size?
Reason is that we wish to preserve the "print size" and resize it based on a dpi value instead.

Example: 
CMYK TIF file (with possible transparency), 300dpi, 3000x3000 pixels = print size 10 inch
If we wish to scale it down to 72 dpi and keep the print size of 10 inch, we should scale it down to 720  pixels x 720 pixels and set the DPI to 72 when saving it. But to be able to do that we would need the info for the original dpi and/or print size to correctly calculate the new pixel dimension of the file to keep the print size.

Can we in libvips read the current dpi of the input file and then do the math . Or is the DPI and/or print size not available as info for a file through libvips?

Would this be a good approach (if possible):

1. Load image using sequential access
2. read properties such as size in pixels, colorspace and hopefully dpi as well
3. close the first opened file access (only used to get the info about the file)
4. perform math to calculate new size
5. Load image again, but this time use thumbnail function (as it combines functions suited for scaling down files) to scale down the file to correct size (as well as icc transformation if needed).
6. Save the image in same or other format.

Or should I not use the "thumbnail" function at all and just load it once, get info about the file, perform math, resize it , icc transform and save it, doing all the stuff "manually" instead of the combined functionality in "thumbnail" (would have to take care of possible transparency then as well)?

Any thoughts?

**Comment by hanssonrickard** [comment]
Stupid me.
I can see now that the xres and yres is available and i can use those to do the calculation.
Then the question that remain is if i should load image and do it manually, or if i should use thumbnail function to do it.

**Comment by jcupitt** [comment]
Hi @hanssonrickard,

Yes, libvips is designed so that opening an image is always quick -- it'll only decode pixels when they are needed. This means you can open, read any fields you need, then open again to process.

In python, something like (untested):

```python
im = pyvips.Image.new_from_file("something.tiff")
target_dpi = 72
scale_factor = im.target_dpi / im.xres
new_width = im.width * scale_factor
resized_image = pyvips.Image.thumbnail("something.tiff", new_width)

# need to update DPI
resized_image = resized_image.copy(xres=target_dpi)
resized_image.write_to_file("new_file.tif")
```

**Reply by hanssonrickard** [reply]
Thanks.
I guess i need to convert the xres to dpi (dots per inch) because according to docs it seems to be dots per mm in pyvips.
Will give it a try.

Thanks for really quick and helpful response.

**Reply by hanssonrickard** [reply]
I had to do some small tweaks, such as that pyvips/libvips report xres in dots per mm, while we usually use dots per inch in print industry.

im = pyvips.Image.new_from_file(input_file, access='sequential')

res_per_mm = (output_resolution / 25.4)
scale_factor = get_scale_factor(in_res=im.xres, out_res=res_per_mm)
new_width = im.width * scale_factor

resized_image = pyvips.Image.thumbnail(input_file, new_width)
resized_image = resized_image.copy(xres=res_per_mm, yres=res_per_mm)

resized_image.write_to_file(output_folder + '/' + file_name)

A few strange things.

1. Opening the resulting file in photoshop, resolution is still reporting 300 dpi, even if the file has info that it is 72 dpi.
If i open the file in Indesign or in macos preview, the reported resolution is correct. So i am unsure why Photoshop reports it wrong. The only difference i can see is that the reportd resolution in the metadata is saved in cm format, if i resave it in photoshop it is saved in inch format (and then photoshop shows it correct). 

2. Layers in the original tiff file is no more. Everything has been flattened, but i guess that is because libvips does not read/handle Tiff layers (as we have discussed in another thread) and therefor (have to?) ignore the Photoshop filen tags 37724. The content is correct, but not split up in the individual layers anymore. 

(Also visible in the log):

2021-04-08 00:03:34,128 [352624] WARNING VIPS: Unknown field with tag 37724 (0x935c) encountered
2021-04-08 00:03:34,130 [352624] WARNING VIPS: Unknown field with tag 37724 (0x935c) encountered
2021-04-08 00:03:34,132 [352624] WARNING VIPS: Unknown field with tag 37724 (0x935c) encountered 

3. This might be the biggest issue. 
If i open the TIFF in photoshop and look at it, all looks good with the transparency, but if i drag and drop it into an Indesign document for a layout, the transparent background shows as black and not transparent. That of course will be an issue if using those TIFF files in a layout. I have made a screenshot of it, where both versions are visible and the libvips tif is showing as black.


I hav attached a zip file with the original tif, the converted tif as well as a screenshot of how it looks in Indesign.
Do you possible know why it might interpret the transparency as black?


[Arkiv.zip](https://github.com/libvips/libvips/files/6274862/Arkiv.zip)

**Comment by jcupitt** [comment]
Ooop you're right, libvips is always pixels/mm everywhere.

**Comment by jcupitt** [comment]
`tiffinfo` is the standard tool for checking the insides of a TIFF file. I see:

```
$ tiffinfo downscaled_transparent_and_layers.tif
TIFF Directory at offset 0x96698 (616088)
  Image Width: 453 Image Length: 340
  Resolution: 28.3465, 28.3465 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: RGB color
  Extra Samples: 1<unassoc-alpha>
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 4
  Rows/Strip: 128
  Planar Configuration: single image plane
  XMLPacket (XMP Metadata):
<?xpacket begin="﻿" id="W5M0MpCehiHzreSzNTczkc9d"?>
... snip
<?xpacket end="w"?>
  Photoshop Data: <present>, 5514 bytes
  ICC Profile: <present>, 3144 bytes
```

So the downsized TIFF file has the correct resolution (28.3 pixels/cm is 72 dpi), but there's a lot of extra metadata in XMP and private photoshop data still there. Photoshop is getting the resolution from this private data, not from the standard TIFF header. The same thing is happening with your transparency: if I load that TIFF into a standard viewer, transparency works. But because Photoshop is also looking at the XMP and private metadata, it gets confused and turns alpha into an extra black layer.

You could try removing this extra metadata. Something like (untested, again):

```
# need to update DPI
resized_image = resized_image.copy(xres=target_dpi)
resized_image.remove("xmp-data")
resized_image.remove("photoshop-data")
```

The underlying problem is that photoshop does not read and write standard TIFF files, unfortunately. Or perhaps, even more abstractly, that TIFF is so flexible that programs are able to create private file formats that only look a bit like standard TIFFs.

**Reply by hanssonrickard** [reply]
Thanks.

As you say, all other places/applications seems to interpret the resolution correct.
What strikes me is how it differs between adobes own applications.

Photoshop: Shows transparency correct, shows dpi wrong.
Indesign: Shows transparency wrong, reads dpi correct

Inconsistent behavior.

Anyhow. I will give it a try to remov parts or all of the metadata and see if that fixes the issue.
Thanks again for quick response.

**Reply by hanssonrickard** [reply]
Removing the photoshop-data fixed the issue with the resolution reported in Photoshop.
But the black background persist in Indesign when mounting the tif file in a indd document.

In that case it did not resolve anythng removing photoshop or xmp data from the file.

**Comment by jcupitt** [comment]
You could try converting to PNG first -- indesign ought to read that correctly.

**Reply by hanssonrickard** [reply]
The case is that the customer works on highres tif files.
While they are being worked on, an external person would like to be able to download a lower resution of that file so they can use it in their indesign documents for placement nad start doing the layout, but they do not wish them to use the highres files (no highres files should circulate outside the house before they are fully ready for print).

Therefor, creating a lower resolution file (based on dpi, so it retains the print size), with same file extension (and possible that we add a watermark to it via libvips) is the ideal solution as they can create their layout with the low resolution tif file and when the highres tif is finished, all they need to do is to "update" the link in indesing (when replaced low res file with highres) without having to relink it to a different filename (filetype).

Therefor, using PNG in this case is not a good solution. We wish to match the filetype between highres/lowres versions.


Back to the problem with transparency, if i just open it in photoshop and resave it, it works fine in Indesign. but we do not wish to do that of course.

I do see a difference between them and that is what is reported via exiftool called "extra samples".
For the tiff from libvips i get a value of "Unassociated alpha", but for the photoshop saved file i get the value "Associated Alpha".


Do you think that can be related to the issue?

I found this by searching about it:


Unassociated alpha data is transparency information that logically exists independently of an image. Unassociated alpha is additional information, which can be removed or edited.

Associated alpha data is opacity information. The color information of each pixel is additionally premultiplied with the value of the alpha channel. If this channel data is removed or edited, the image changes. Storing data in premultiplied format allows compositing operations to be implemented more efficiently.


Is this something that we can affect using libvips (at least for a test to see if that causes the issue)?

**Reply by hanssonrickard** [reply]
Additional info, the original highres has the value "Associated alpha" for "extra samples" field.

**Comment by jcupitt** [comment]
Yes, that could be it, but indesign is breaking spec there. Unassociated alpha is still alpha, just not premultiplied.

libvips undoes premultiplication in associated alpha channels on tiff load, but always saves as unpremultiplied alpha. We'd need to add an extra flag to the tiff saver to do premultiplication on save.

In the short term, I think your best option is to use PNG for these low-res files.

**Comment by jcupitt** [comment]
I made an issue for this enhancement: https://github.com/libvips/libvips/issues/2192
